### ISSUE TITLE
Failed executions do not indicate successful changes

### ISSUE THREAD

**Issue by aprice** [hashicorp/terraform#17681]
### Terraform Version
```
Terraform v0.11.4
+ provider.aws v0.1.4
+ provider.template v0.1.1
```

### Expected Behavior
When an error occurs during execution of a template, error output should indicate what changes were successfully executed before the failure occurred, so that (given the clearly indicated `Terraform does not automatically rollback in the face of errors.`), an operator can quickly address any partial changes if necessary.

### Actual Behavior
Error output indicates which resource(s) failed, and that they will not be rolled back, but not what changes had been committed up to that point (i.e. what an operator might want to manually roll back before further troubleshooting their Terraform configuration).

### Additional Context
This can be particularly confusing/misleading when resources encompass many API calls to effect changes. For example, a change to an `aws_security_group` resource actually results in a series of API calls to remove and add various rules from/to the group. If one of those calls fails, all the operator knows is that the `aws_security_group` resource had an error - but, critically, _not what changes have been made to the group_. The failure could range from innocuous (no changes committed before failure) to critical (removes executed but adds failed. leaving no rules and denying all traffic). Depending on the context, an operator might need to execute a rapid manual intervention before further troubleshooting within Terraform for a more permanent fix; but they can't make a good call if they don't have sufficient information on what Terraform did and didn't do.

**Comment by nbering** [comment]
This is a good idea... I think it's a great to provide as much information as is possible when a failure occurs.

That said... the specifics of what you're asking for might be difficult to achieve. Resources are heavily abstracted from core, so if a provider makes many calls to set up a resource, Core doesn't know anything at all about that. It just gets an error message back if it fails, and/or some state data which conforms to the structure you see in TFState, and is documented on terraform.io.

It might be worth exploring some ways for providers give more helpful logs back out to the user, though. So far, providers have a pretty limited ability to provide user feedback, unless you're in debug mode.

**Comment by aprice** [comment]
I'm not well-versed in Terraform's internals, but wouldn't this be possible based on the state change indicated by the provider? If not always, at least in the case of "partial mode"?

> Partial mode is a mode that can be enabled by a callback that tells Terraform that it is possible for partial state to occur. When this mode is enabled, the provider must explicitly tell Terraform what is safe to persist and what is not.

Or is it that providers that error mid-execution are not updating the state when this happens? That seems like it could also be an issue, though only tangentially related to user-visible output.

Debug mode is handy but unless you make a habit of _always_ running in debug mode, by the time you need it (for the type of situation I'm describing here)m it's too late.

**Comment by nbering** [comment]
You are correct, I had forgotten about this particular detail in describing (as I've experienced it) the interface between Terraform and a provider. As a Providers Developer, generally all we get to do is tell Terraform which variables cause are computed, which cause the resource to be destroyed, etc. And then when a resource needs to change, the API presents an interface for updating the state, and returning and error, etc.

There's nothing in the interface to say, "Tell the user this thing happened." As far as I am aware, all we get is debug logging, or return an error.

**Comment by nbering** [comment]
Sorry... I come off sounding confrontational sometimes and I don't mean to. I think there is an opportunity to improve here, but I'm just pointing out a constraint in the current design of the system that makes this more challenging than a simple change to the text format of the output on an error, or something like that.

I'm curious if there's anything like this already being discussed in other issues, and how this might fit into the current internal roadmap. The team usually focuses on one aspect at a time for each breaking release... and I'd bet this would need a breaking release to fit in.

**Comment by jbardin** [comment]
Hi @aprice,

Thanks for opening the issue. There might be something that Terraform could do better in this case in the UI, but first I think we would need to decide what that would look like before we could make a plan. 

In most cases, providers apply a single resource at a time (though in the less-common case they have multiple steps, they can use partial state application), so if a resource fails, then it is _only_ that resource that fails, which will prevent any dependent resources from being applied. 

The usual way to see what was applied, is to run a plan after to see what changes are remaining. Terraform could track the changes applied, and report a summary after an error, which in theory should be the compliment of the next plan.

The case you mention, where there are multiple calls for a single resource, and some intermediate step failed is hard for terraform to report in the state, because the state does not track api calls, just the state of resources. My more immediate plan for this is simply to improve the log handling throughout the code, so that out of band warnings and errors can be properly reported to the user. As for _what_ to report to the user, I think is very much on a case-by-case basis depending on the provider. If the series of calls is not idempotent, and can't be applied in some sort of transaction, then it is the provider's responsibility to handle possible inconsistencies.

**Comment by aprice** [comment]
I'm curious - you say "the state does not track api calls, just the state of resources." Does that include partial resource changes in the event of an error? In the example given with an AWS security group, would the new state reflect that the security group now has no rules? If so, it seems like that state change, if it were reported to the user in the event of an error, would be enough information to deliver to the operator to know what had happened and make a tactical decision at that point. More logging would certainly be beneficial but I understand that would require changing each resource handler in each provider which could be a monumental task.

**Comment by nbering** [comment]
That would depend how the Provider handles the error. Assuming they've handled partial state, and done so correctly, the state should reflect reality.

There are some cases, when dealing with errors, that it might not be so clean. For example, when changing the size of an instance, you need to stop the instance first.

Instance state is not tracked, so if the instance is stopped, and the resize results in an error... a change has now been made that would not be recorded in the partial state update.

It's a bit of a contrived example - depending on something I personally already considered to be a design flaw in the provider implementation - but I think it represents the disconnect between the specific case you mentioned (multiple API calls in an action), and using the implementation of partial state to report side-effects from a partially-complete apply.

**Comment by lyda** [comment]
A possible use case for this:

A long while ago I worked at a place we'll call the Chocolate Factory. We had a tool like terraform (let's call it Smasher) that would spin up large clusters and generally manage the state and workflow of this sort of work. One thing it did was to file tickets when it needed a human to do things.

Once the ticket was addressed and closed successfully, Smasher would continue running. I'm not quite sure how this might work with terraform since it could end up in states where it filed the ticket and would then need to stop in an unfinished, but complete for now state.

Imagine this scenario: you work at a place where Sally has to sign off on EC2 instance creation.  This works by filing a jira ticket assigned to Sally and then she'll either approve of reject the ticket. You could script this into terraform if you could do partial status updates. The initial run would create a ticket and then exit in a failed but still processing state with the ticket info saved in the state file.  You could run a few times more with the same result but no new ticket filed.

When Sally approves the ticket, the next terraform run would see that and continue with creating the EC2 instances. If she rejects, it will have a permanent failure and at that point it will never succeed and then politics start - which is clearly outside the scope of terraform bar just noting that's where we're at.

But partial state saves would be very useful here to make this work.

**Comment by apparentlymart** [comment]
Hi @lyda,

The Terraform "plan" feature is Terraform's intended solution to that use-case. Rather than having the asynchronous human review happen while _applying_ the changes, instead we ask Terraform to produce a description of what it intends to do and review that before any changes are made at all. Once humans are all satisfied that the plan is acceptable, Terraform can then non-interactively apply those changes.

In more complex scenarios, it is possible to have an automaton run `terraform plan -out=tfplan` and then `terraform show -json tfplan` to get a machine-readable representation of the plan, which could then be used to make decisions. In your case for example, perhaps if there's any plan involving an `aws_instance` or `aws_autoscaling_group` resource then it would tag Sally for review before the change is even committed to the main branch in version control.

I understood this issue as being more about what happens when things _don't_ go to plan: everything is approved and expected to work, but some assumption didn't hold and so one of the apply actions fails. In that case, you can refer back to the verbose logs above the error to see what Terraform has done so far, but I think the request was to repeat that information in a more concise format as part of the error message.

Although the error message talks about "rolling back", I think in practice most users of Terraform don't actually do that: instead, they make a _new_ change to fix the error and "roll forward" to that change.

In situations where you _do_ want to roll back, reverting the configuration back to how it was when the previous change was applied will often achieve that, though there are some changes where that isn't sufficient, such as when a whole new provider was introduced and so rolling back requires destroying infrastructure using the provider configuration that was added at the same time as the new resources were added.

Generally though, when we talk about "rolling back" in a Terraform context we're talking about undoing changes to the configuration, not undoing changes to the infrastructure itself. With a few edge-case exceptions, the Terraform model calls for any change to infrastructure to be represented by a change to the code that describes it, and that includes "rolling back": the reversion back to a previous state should ideally be implemented by resetting the configuration back to the previous state and running Terraform again, so that the version control changes always precede the real infrastructure changes they imply.
